core/embed/rust/src/crypto/aesgcm.rs

```diff
@@ -0,0 +1,356 @@
+use core::{mem::MaybeUninit, pin::Pin};
+
+use zeroize::{zeroize_flat_type, Zeroize};
+
+use super::{ffi, Error};
+
+type Memory = ffi::gcm_ctx;
+
+impl Default for Memory {
+    fn default() -> Self {
+        // SAFETY: a zeroed block of memory is a valid aes_gcm
+        unsafe { MaybeUninit::<Memory>::zeroed().assume_init() }
+    }
+}
+
+// Can't use DefaultIsZeroes as we don't have Copy.
+impl Zeroize for Memory {
+    fn zeroize(&mut self) {
+        // SAFETY:
+        // - gcm_ctx does not contain references to outside data or dynamically sized
+        //   data
+        // - values do not have Drop impls
+        // - can invalidate the type if it is used after this function is called on it -
+        //   only used in Drop
+        // - all zero bit pattern is valid context, see Default impl
+        unsafe { zeroize_flat_type(self as *mut Self) };
+    }
+}
+
+// Tag size is a parameter but we fix it to 16 here for simplicity.
+pub const TAG_SIZE: usize = 16;
+pub type Tag = [u8; TAG_SIZE];
+
+// for bindgen RETURN_* macros are u32, just redefine the only one we are using
+const RETURN_GOOD: i32 = 0;
+const KEY_SIZES: [usize; 3] = [16, 24, 32];
+
+#[repr(u8)]
+#[derive(PartialEq)]
+enum State {
+    Init,
+    Encrypting,
+    Decrypting,
+    Finished,
+    Failed,
+}
+
+pub struct AesGcm<'a> {
+    ctx: Pin<&'a mut Memory>,
+    state: State,
+}
+
+impl<'a> AesGcm<'a> {
+    pub fn new(mut ctx: Pin<&'a mut Memory>, key: &[u8], iv: &[u8]) -> Result<Self, Error> {
+        if !KEY_SIZES.contains(&key.len()) {
+            return Err(Error::InvalidParams);
+        }
+        // initialize the context
+        // SAFETY: ffi
+        let res = unsafe {
+            ffi::gcm_init_and_key(
+                key.as_ptr(),
+                key.len() as _,
+                ctx.as_mut().get_unchecked_mut(),
+            )
+        };
+        if res != RETURN_GOOD {
+            return Err(Error::RuntimeError);
```

Comment:
Maybe define an AES-GCM specific error variant?

This way we can return a bit more detailed error code:
```suggestion
            return Err(Error::AesGcm(AesGcmError::InitAndKeyFailed));
```

Reply:
Not sure if such specific errors are worth it, not much we can do in these "something failed that shouldn't fail" cases.

Anyway I reworked the error handling a bit in 53a0a522eb0fcf65f40ba62a130684aff2016127. Most of the `gcm_...` C functions always return `RETURN_GOOD` so these are `ensure!`d and not propagated further. Changed `RuntimeError` to `InvalidContext` but open for suggestions.